packages/terra-application/src/primary-navigation-layout/workspace-layout/WorkspaceLayout.jsx

```diff
@@ -0,0 +1,458 @@
+};
+
+const getActiveSizeForWorkspaceSize = workspaceSize => {
+  if (workspaceSize.scale === 0) {
+    return 'small';
+  } if (workspaceSize.scale === 0.5) {
+    return 'medium';
+  } if (workspaceSize.scale === 1) {
+    return 'large';
+  } if (workspaceSize.type === 'split') {
+    return 'split';
+  } if (workspaceSize.type === 'overlay') {
+    return 'overlay';
+  }
+
+  return undefined;
+};
+
+const validateInitialWorkspaceSizeForBreakpoint = (breakpoint) => {
+  if (breakpoint === 'large' || breakpoint === 'huge' || breakpoint === 'enormous') {
+    return true;
+  }
+
+  return false;
+};
+
+const WorkspaceLayout = ({
+  children,
+  id,
+  workspace,
+  contentElementRef,
+}) => {
+  const activeBreakpoint = React.useContext(ActiveBreakpointContext);
+  const parentLayoutActions = React.useContext(LayoutActionsContext);
+  const applicationIntl = React.useContext(ApplicationIntlContext);
+
+  const pageContainerRef = React.useRef();
+  const layoutBodyRef = React.useRef();
+  const workspacePanelRef = React.useRef();
+  const workspaceResizeBoundsRef = React.useRef();
+  const resizeOverlayRef = React.useRef();
+  const workspaceOverlayRef = React.useRef();
+  const lastActiveSizeRef = React.useRef();
+  const userSelectedTypeRef = React.useRef();
+  const userSelectedScaleRef = React.useRef(0);
+
+  let initialWorkspaceSize;
+  if (validateInitialWorkspaceSizeForBreakpoint(activeBreakpoint)) {
+    initialWorkspaceSize = (workspace && workspace.props.initialSize) || initialSizeForBreakpoint(activeBreakpoint);
+  } else {
+    initialWorkspaceSize = initialSizeForBreakpoint(activeBreakpoint);
+  }
+
+  const [workspaceSize, setWorkspaceSize] = React.useState(initialWorkspaceSize);
+  const hasOverlayWorkspace = activeBreakpoint === 'tiny' || activeBreakpoint === 'small' || workspaceSize.type === 'overlay';
+  const isLargeFormFactor = activeBreakpoint === 'large' || activeBreakpoint === 'huge' || activeBreakpoint === 'enormous';
+  const [workspaceIsVisible, setWorkspaceIsVisible] = React.useState(!hasOverlayWorkspace && workspace && workspace.props.initialIsOpen);
+
+  const layoutActionsContextValue = React.useMemo(() => {
+    let newActions = parentLayoutActions.actions || [];
+
+    let actionLabel;
+    if (workspaceIsVisible) {
+      actionLabel = applicationIntl.formatMessage({ id: 'terraApplication.workspaceLayout.toggle.hide' });
+    } else {
+      actionLabel = applicationIntl.formatMessage({ id: 'terraApplication.workspaceLayout.toggle.show' });
+    }
+
+    if (workspace) {
+      newActions = [...newActions, {
+        key: 'workspace-layout-toggle-workspace-panel',
+        label: actionLabel,
+        icon: workspaceIsVisible ? <IconPanelRight /> : <IconPanelLeft />,
+        onSelect: () => {
+          setWorkspaceIsVisible(state => !state);
+        },
+      }];
+    }
+
+    return ({
+      actions: newActions,
+    });
+  }, [parentLayoutActions.actions, workspace, workspaceIsVisible, applicationIntl]);
```

Comment:
You could stick a breakpoint in here to see how ofter this effect fires. The 'workspace' value might be triggering this effect more often than it needs to.

A way around this would be to do something like:
```jsx
  const hasWorkspace = !!workspace; 

  const layoutActionsContextValue = React.useMemo(() => {
    ...
    if (hasWorkspace) {
      ...
    }
    ...
  }, [parentLayoutActions.actions, hasWorkspace, workspaceIsVisible, applicationIntl]);
```
to make the dependency array more representative of what you care about here.

Reply:
Updated.